From 96971fb25248f66c1e8a4fd7a0a4b75c4b67674a Mon Sep 17 00:00:00 2001
From: Mostafa Saleh <smostafa@google.com>
Date: Thu, 16 Apr 2026 11:35:51 +0000
Subject: [PATCH 123/223] ANDROID: iommu/arm-smmu-v3-kvm: Fix CD size in
 smmu_dev_block_dma()

The cdtable size should be the nr_entries multiplied by entry size
where the nr_entries is 2 ^ pasid_bits.

However, smmu_dev_block_dma() raises the nr_entries to the power of 2
which is wrong.

Bug: 503376736
Change-Id: Ica4786866f04c421819bc983e824232922cafcc8
Signed-off-by: Mostafa Saleh <smostafa@google.com>
(cherry picked from commit 7d6d4a26e7ead0bc501f3ed6809a66c2f3d49365)
Signed-off-by: Lee Jones <joneslee@google.com>
---
 drivers/iommu/arm/arm-smmu-v3/pkvm/arm-smmu-v3.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/iommu/arm/arm-smmu-v3/pkvm/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/pkvm/arm-smmu-v3.c
index f13a435b763a..fb08decd7784 100644
--- a/drivers/iommu/arm/arm-smmu-v3/pkvm/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/pkvm/arm-smmu-v3.c
@@ -1652,7 +1652,7 @@ static int smmu_dev_block_dma(struct kvm_hyp_iommu *iommu, u32 sid, bool is_host
 							    STRTAB_STE_0_S1CTXPTR_MASK);
 				nr_entries = 1 << FIELD_GET(STRTAB_STE_0_S1CDMAX,
 							    le64_to_cpu(dst->data[0]));
-				cd_sz = (1 << nr_entries) * (CTXDESC_CD_DWORDS << 3);
+				cd_sz = nr_entries * (CTXDESC_CD_DWORDS << 3);
 			}
 			/* zap zippity zop. */
 			for (i = 0; i < STRTAB_STE_DWORDS; i++)
-- 
2.50.1

